### PR TITLE
[WIP] Optimize Steady State Adjoint

### PR DESCRIPTION
**Pull request by avik-pal**
Currently, VecJac operator repeatedly call the forward pass with the same arguments.
(I know this is not the right place to keep the operators and stuff. But wanted to share a temporary solution which should speed things up for big neural networks)

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
We should define an operator in this package with all of the vjp choices.

File: src/steadystate_adjoint.jl [hunk 1]
```diff
@@ -21,88 +14,38 @@ struct SteadyStateAdjointSensitivityFunction{
...
-        sol,
-        dg,
-        f;
-        quad = false,
-        needs_jac = needs_jac,
-    )
+    diffcache, y = adjointdiffcache(g, sensealg, discrete, sol, dg, f; quad=false, needs_jac=needs_jac)
 
     λ = zero(y)
     linsolve = needs_jac ? nothing : sensealg.linsolve
```
Comment (ChrisRackauckas) [inline src/steadystate_adjoint.jl:23]: it should always just use `sensealg.linsolve` and default that to `nothing`.

File: src/steadystate_adjoint.jl [hunk 2]
```diff
@@ -21,88 +14,38 @@ struct SteadyStateAdjointSensitivityFunction{
...
-        discrete,
-        sol,
-        dg,
-        f.colorvec,
-        needs_jac,
-    )
+    sense = SteadyStateAdjointSensitivityFunction(g, sensealg, discrete, sol, dg, f.colorvec, needs_jac)
     @unpack diffcache, y, sol, λ, vjp, linsolve = sense
 
     if needs_jac
```
Comment (ChrisRackauckas) [inline src/steadystate_adjoint.jl:43]: `needs_jac = linsolve !== nothing && needs_concrete_A(linsolve)` using https://github.com/SciML/LinearSolve.jl/pull/82

**Comment by avik-pal** [comment]
Not modifying this branch for the sake of the paper. New PR #542

**Pull request closed without merging**